### ISSUE TITLE
Wrong exit number at roundabout

### ISSUE THREAD

**Issue by pl71** [osmandapp/OsmAnd#19796]
### Describe the routing engine used (required)

- [ X] OsmAnd's in-app offline routing
- [ ] Any online routing provider (YOURS, openrouteservice, OSRM, etc.)

### Describe your routing Profile (required)

Car

### Describe your start and end points (required)

https://osmand.net/map/?start=44.264825,26.046985&end=44.266438,26.054270&profile=car#17/44.26680/26.05088

Same behavior observed on several other roundabouts.

### Describe the actual route (required)

Route is correct. I have wrong next turn instruction on OsmAnd.
OsmAnd is showing exit number 1 instead of 2.
<table style="width:100%">
  <tr>
<th><img src="https://github.com/osmandapp/OsmAnd/assets/4776779/ea7b8e41-640a-4cf8-961f-13c2f4c4c392"  width="300" height="auto"/> </th>
  </tr>
</table>

Online engine `osmand.net` is reporting exit number 2 correctly.
When simulating, turn instruction is correct also:

<table style="width:100%">
  <tr>
<th><img src="https://github.com/osmandapp/OsmAnd/assets/4776779/13c999ec-9f25-440c-ab74-1054b74b5faf"  width="300" height="auto"/> </th>
  </tr>
</table>

### Describe the expected route (required)

Next turn instruction is expected to be "Exit 2" at the roundabout.

### Describe what Maps you used (online or offline) (required)

- [X] Offline maps offered within the OsmAnd app for download. - Romania 1.05.2024
- [ ] Online (tile / raster) maps 

### Your Environment (required)

```
OsmAnd Version: 4.7.10
Android/iOS version: 14
Device model: S22 Ultra
```

### Anything else? (optional)
No 1 exit is shown wrongly on several other roundabouts during the same trip:
/Possible software bug/

<table style="width:100%">
  <tr>
<th><img src="https://github.com/osmandapp/OsmAnd/assets/4776779/114dd4c0-7f3d-4b44-9bce-a7f47d56df2e"  width="300" height="auto"/> </th>
 <th><img src="https://github.com/osmandapp/OsmAnd/assets/4776779/240fe877-db01-494f-af23-cad1714f5175"  width="300" height="auto"/> </th>
  </tr>
</table>

**Comment by pl71** [comment]
The whole trip was with weird and sometimes missing instructions. You may close the issue since no other users complain here.

**Comment by EtlamGit** [comment]
I have the same issue and it occurs at least since last year.
And it is reproducible when navigating the same route.

All roundabouts in e.g. Trollhättan have wrong exit numbers. But other roundabouts on the same trip show correct number.
Example: https://www.openstreetmap.org/#map=17/58.28030/12.30540

What I do to reproduce it:
* start in Göteborg, Sweden
* set a destination far north from Trollhättan
* drive that track !

you will notice wrong exit numbers on all major streets and correct ones on smaller ones.
**The original bug reporter stated that it is not happening with simulation. So you have to drive yourself!!!**

I have tried to find differences in underlying OSM raw data since two years now, and the only thing I found is:
It seems that it happens when different categories of roads are connected to the same roundabout. When I approach on the higher category road and want to exit on the same category, I get "exit 1" even when there are lower categories of roads before that in the roundabout.
![grafik](https://github.com/osmandapp/OsmAnd/assets/5821779/79dd61f2-ec54-417c-ba21-4410dd42fffb)


Since two or three years I'm not able to pass Trollhättan without this bug occurring.
I always use the latest OsmAnd version and offline map version that is available 1 week before start of the trip.

**Comment by EtlamGit** [comment]
@vshcherb what to do to get this reopened and solved?

I can offer to test this route around Trollhättan twice each year.

**Comment by EtlamGit** [comment]
@yuriiurshuliak did you drive that track yourself? Did you navigate a longer distance?

The original bug reporter stated that it is not happening with simulation. I can confirm that behavior: error in real navigation, all fine in simulation. You have to drive yourself to trigger it!

**Comment by yuriiurshuliak** [comment]
I reviewed the roundabout you mentioned and found that the exit numbers were displayed correctly. However, our testing was limited to navigation simulation.

For the steps you provided, you mentioned a route from Göteborg, Sweden, to an area north of Trollhättan, which takes about two hours. It would be helpful if you could specify the exact locations where the incorrect exit numbers appear.

Additionally, please note that exit numbers are counted only on permitted roads. If a road is restricted (e.g., pedestrian-only or brick access), it won't be included in the exit count.

OsmAnd~ 4.8.0#2722m, released: 2024-07-12

<table style="width:100%">
  <tr>
<th><img src="https://github.com/user-attachments/assets/1bc9418b-853f-4d6b-a6e0-f89cb2bc010f"  width="300" height="auto"/> </th>
  </tr>
</table>

**Comment by EtlamGit** [comment]
(Just coming back from a trip to Stockholm with correct roundabout numbers)

Yes, for me it is the same result, when I try to simulate only short routes around these roundabouts, everything is fine.
But each time I do it in real life, it fails.

None of the other roads is restricted.

All roundabouts along E45 in Trollhättan are affected.

More information about my typical trip:
I start in Göteborg harbor: https://www.openstreetmap.org/#map=16/57.6957/11.9143
At start of route there is no GPS signal, also in Göteborg GPS is lost in tunnels. I do not follow the navigation instruction because it leads to crowded city roads. Instead I turn West and take this large bridge. After some time OsmAnd has calculated several times a new route but sometimes also states "45000km left". So at start there is is quite much to calculate, maybe also much RAM used.
First stop and destination of navigation is e.g. Säffle: https://www.openstreetmap.org/#map=19/59.13528/12.91812

I typically do voice notes while traveling (to fix stuff later). For this year I have notes about wrong roundabouts with the following file names. (I'm sure you are are an experienced OSM user and able to use these to get locations):
0YTJdcMf--.1.3gp
0YTMGmlQ--.1.3gp
0YTbPiA1--.1.3gp
0YTbyKSO--.1.3gp
0YT0BQe7--.1.3gp
0YT0SOV6--.1.3gp
0YT0YT85--.1.3gp
Obviously the location will be only near the roundabout, as it is hard to tap-talk and drive at the same time. 


Is there some debug log I could generate next time I do this trip?

**Comment by EtlamGit** [comment]
At the moment I'm using the official last build from Google Play Store (OsmAnd+ 4.7.17). As I have limited internet access and especially limited amount of data volume, I'm not that eager to test other versions.
If there is a simple APK available, I could try to setup a spare device and let them run in parallel.

My next passing to these locations will be end of July. So we have some time left to prepare debugging/logging.

**Comment by mosiser** [comment]
Same problem.
Osmand 4.7.17 from f-droid.
https://osmand.net/map/navigate/?start=45.939774,12.322599&end=45.949516,12.323742&profile=car#17/45.94443/12.32153
In both roundabouts Osmand tells me to take exit 1 during real navigation.
Instead, it correctly reports exit 2 during simulation.

Maybe something to do with the fact that the ignored roads are of lower classification than the one I'm following?

**Comment by EtlamGit** [comment]
I have prepared multiple devices for the next trip (official 4.7.17 and 4.8 nightly from 2024-07-22)
It is not guaranteed to happen again, but lets try . . .

@DmitryAlexei How do I enable debug logging?

**Comment by pl71** [comment]
My trip was relatively long - abt. 300 km and crossing national border. It means using 2 maps. The bug appeared on the second map.

**Comment by EtlamGit** [comment]
Here is my feedback from the last trip:
(TLDR: the latest version 4.8 works)


Prepared multiple devices to compare low/high memory and app version:
* old Tablet that I have mounted in car with 4.7.17
* old Smartphone with 4.8 nightly from 2024-07-22
* high end smartphone with 4.7.17

All devices had the same map data files from end of June.

Did the above mentioned route in reverse direction (Säffle via Trollhättan to Göteborg harbor):
* restarted all devices before start
* let all devices find GPS signal
* set the same target and started navigation

Result:
* the new version 4.8 reported correct navigation guides all the time
* the old version 4.7.17 failed in several aspects (not only roundabouts):
  * right from the start there was a very long distance without any advices (~150km straight) whereas the new version has some additional turn advices
  0ZKX7nlH--.1.3gp  missing turn advise
  0ZIwRScz--.1.3gp  missing turn advise
  * starting with the first roundabout it always instructed to use first exit instead of second:
  0ZI23RGw--.1.3gp
  0YdVWzD6--.1.3gp
  0YW32aDA--.1.3gp
  0YW3x6rA--.1.3gp
  *  the next roundabout was already announced with "take first" when I took an unplanned exit to buy some food and gas for the car. This took some time and I drove several circles in an industrial/shopping district (also crossing ~5 smaller roundabouts).
  * **when coming back to the planned route the next roundabouts were announced correctly from all devices!** That was different in last year.
  0YT0VXBk--.1.3gp (this one was visible in wrong state before !!!)
  0YT0SnYd--.1.3gp
  0YT0BUp~--.1.3gp
  0YTbyLGM--.1.3gp
  * but after some time it started to fail again:
  0YTblEt_--.1.3gp
  0YTMGk9n--.1.3gp
  * until the last roundabout where the advice "take first" was correct as there is no other exit before
  0YTMIDi1--.1.3gp

It might be fixed in version 4.8
But I do not understand why this instable behavior did not pop up in some regression test for more than 1 year. There are several other bugs in OSMand that come back every other year, so I would highly appreciate if someone will:
* understand the bug reason in the old version
* write some regression test to prevent this from coming back !

**Comment by DmitryAlexei** [comment]
@EtlamGit thank you very much for your feedback
Please note, that we are constantly working on improving turns indication. In 4.8 several bugs have been fixed. 
It also appears that some users encounter incorrect turn indications after multiple route recalculations. This behavior is hardly reproducible.
